### ISSUE TITLE
Breadcrumbs failure in Tunnel section of Finals Qualification world

### ISSUE THREAD

**Issue by m3d** [osrf/subt#810]
I am observing strange behavior of breadcrumbs in tunnel section of finals qualification world (going straight on the first crossing). Yesterday I run experimentally 4x Freyja robots (robotika, `ver99octofqc`, `a0ac3e78-b0a8-4b41-9567-637f2de4a41e`, one to left/Urban, one to right/Cave, and two straight/Tunnel), and if you see report from Teambase how are they went, you can tell, that in tunnel is something wrong:
![finals-breadcrumbs](https://user-images.githubusercontent.com/5664353/110269985-2589ba00-7fc5-11eb-855f-d61c52776589.png)
The Cave robot turned up-side-down shortly after entering the section, so ignore the green. But note the difference of orange robot in Urban part deploying only 6 breadcrumbs and blue/red trace of tunnel (with up-to 12 breadcrumbs from two robots).

The D-robot went all the way to Survivor and Drill, but on the return fall off the vertical shaft as A-robot. The reports of artifact and positions is sent regularly.

I think I am seeing similar problem reported to `subt-help` (Nov 27, 2020) in "Tunnel Circuit World 4" where the breadcrumbs were not helping in long straight tunnel??

**Comment by m3d** [comment]
Here is older run of two Freyja robots exploring the central part and returned successfully home (and thus Teambase has complete trajectory and artifacts):
![two-freyjas](https://user-images.githubusercontent.com/5664353/110271514-84046780-7fc8-11eb-8bd0-0cc9ccbb3a0c.png)

The density of breadcrumbs you can see on video
https://youtu.be/TuSeRfr-DaU

Please let me know what info could help to debug this issue?

**Comment by m3d** [comment]
p.s. I wonder if this could be related to "mix of breadcrumbs from different robots"? Now I am running 3x Freyja (left, center, right) and it works fine and scores as it sees new artifact:
![finals-score3pts](https://user-images.githubusercontent.com/5664353/110304822-520cf880-7ffc-11eb-8c2a-b149dbce2781.jpg)
Also the position of Teambase is slightly different (I think I used the 1st slot this time) and robot could drop breadcrumbs +/- 1m in different location??

**Comment by m3d** [comment]
one more example without breadcrumbs (X4 drone)
![drone-teambase](https://user-images.githubusercontent.com/5664353/110357884-1134d400-803c-11eb-8ee6-fc98dd1687f9.png)
so it looks like the signal ends in the bend ... so in the 4-robots run it was similar to "no breadcrumbs" situation even there were some deployed. Note that "red trace" in the original example was also short (like for the drone) and blue was slightly longer (the first deployed breadcrumb corresponded to "blue robot")??

**Comment by m3d** [comment]
I tried to run two independent runs where two robots enter the Tunnel section and so far only the first one (the one which deployed the breadcrumb in the first turn) is in touch with Teambase all the time ...

Teambase info:
![bread-debug-Tview](https://user-images.githubusercontent.com/5664353/110624655-cd0e1480-819e-11eb-87a5-8d8e0eb212e3.png)

A-robot (the first one, dropping the "common" breadcrumb)
![bread-debug-Aview](https://user-images.githubusercontent.com/5664353/110624764-ed3dd380-819e-11eb-8038-e53e7aa49b1f.png)

B-robot, using shared breadcrumbs and loosing contact with Teambase
![bread-debug-Bview](https://user-images.githubusercontent.com/5664353/110624829-021a6700-819f-11eb-96a1-1c270cc97767.png)

p.s. I may try to create new images and reduce the shared traces to only "online" packets reported by robots at their current position.

**Comment by acschang** [comment]
> p.s. I wonder if this could be related to "mix of breadcrumbs from different robots"?
Every breadcrumb dropped affects the communication ability of every robot (and every robot type) so dropping breadcrumbs from different vehicles in a heterogeneous swarm should not have a negative effect or limit communication in any way.

We appreciate you looking into this issue and it is apparent from your results and our internal testing that the level of communication possible in Tunnel Circuit tiles is not on par with communication in Urban and Cave Circuit tiles.  We will pursue avenues for bridging the communication discrepancy between the three subdomains.

**Comment by m3d** [comment]
@acschang thanks - there could be some other reason, so I am still not sure about the reason. Below are two samples and it is strange that one robot does not drop breadcrumbs
![state tlog](https://user-images.githubusercontent.com/5664353/110687918-4fb7c380-81e1-11eb-8fb3-cd93dc53b90f.png)
![state tlog](https://user-images.githubusercontent.com/5664353/110687937-56463b00-81e1-11eb-9ef1-2a1ea5ba72c0.png)

**Comment by acschang** [comment]
> @acschang thanks - there could be some other reason, so I am still not sure about the root cause. Below are two samples and it is strange that one robot does not drop breadcrumbs

The case you're reporting is definitely strange however we have had instances of this reported before (#655).  We're working on addressing the issue and adding an avenue for breadcrumb deployment feedback (#814).

**Comment by caguero** [comment]
@m3d , any chance you could try [this branch](https://github.com/osrf/subt/compare/issue_810) and reproduce your tests to see if you observe better results in the tunnel section? Thanks!

**Comment by m3d** [comment]
@caguero - thanks for the experiment, unfortunately locally I am not able to run two robots with teambase yet and on CloudSim I cannot choose the branch, so the probability of test in near future is low, sorry :( [I will discuss this with my colleagues]

**Comment by caguero** [comment]
I did some experiments with the default version (not with the suggested branch for testing) and the results look reasonable to me. See the animated gif attached. 

![fq_comms](https://user-images.githubusercontent.com/1440739/113347382-64e1c700-9335-11eb-83c3-d307a914a0c9.gif)

The initial conditions are two robots on the staging area. The first visualization shows the area covered without deploying any breadcrumbs. 

Then, I move forward one robot, pass the tunnel intersection and deploy one breadcrumb in the middle section of the straight tile. The second visualization shows how the comms range extends to one extra tile. 

Finally, I keep moving the robot until the next intersection and deploy another breadcrumb. The comms range extendes a few extra tiles. Note that the comms visualization is always from the perspective of the robot sitting in the staging area.

Could you try to reproduce my experiment? I can also try to reproduce your setup if you provide a minimum experiment that seems strange (ideally with two robots and the poses where you deployed the breadcrumbs).

**Comment by m3d** [comment]
@caguero thanks for the test and nice presentation - it looked to me that the problem is really combination of TWO robots where some breadcrumbs should be shared. In our implementation we share global positions of all deployed breadcrumbs and place them spatially (30 meters spacing at the moment). So maybe I could add logging of strength or what you would suggest? It looks like that only one of the robot (the one which placed the nearest breadcrumb in the first turn of tunnel section) is later able to talk to the Teambase.

**Comment by caguero** [comment]
Could you post a sketch of where the two robots are and where the breadcrumbs are located?

**Comment by m3d** [comment]
> Could you post a sketch of where the two robots are and where the breadcrumbs are located?

probably the best is the "ground truth" from https://github.com/osrf/subt/issues/810#issuecomment-795980251

**Comment by caguero** [comment]
Do you have the poses of the breadcrumbs and robots by any chance?

**Comment by caguero** [comment]
> > Could you post a sketch of where the two robots are and where the breadcrumbs are located?
> 
> probably the best is the "ground truth" from [#810 (comment)](https://github.com/osrf/subt/issues/810#issuecomment-795980251)
> (breadcrumbs are the pink triangles)

And how are the robots represented in the diagram?

**Comment by zwn** [comment]
Breadcumbs are the pink downward facing triangles. There are 2 robots represented by their trajectories and the trajectories are the same color as their names in the legend down left. It is a 2D rendering of state.tlog file after the run finishes.

**Comment by m3d** [comment]
Uhh, it has been a while ... the logs are on CloudSim, just to find the relevant - I will try to look it up. The robots are represented by their colored trajectory.

**Comment by m3d** [comment]
> Do you have the poses of the breadcrumbs and robots by any chance?

The first simulation was `ver99octofqr1-81d7812a-7a4a-47cd-b7f8-b8f47c0a4370` with
```
A210LXM__breadcrumb___0: 29.54 0.57 2.90
A210LXM__breadcrumb___1: 51.68 -19.57 5.00
A210LXM__breadcrumb___2: 81.66 -18.82 5.00
A210LXM__breadcrumb___3: 104.50 0.65 5.00
A210LXM__breadcrumb___4: 134.22 0.62 0.67
```

The second simulation was `ver99octofqbread1-3e9df022-e221-45de-8a5f-a67f8821920f` with
```
B120CXM__breadcrumb___0: 29.54 -0.18 2.90
B120CXM__breadcrumb___1: 48.18 -23.50 5.00
B120CXM__breadcrumb___2: 47.93 -43.55 0.00
B120CXM__breadcrumb___3: 47.81 -49.98 0.00
B120CXM__breadcrumb___4: 48.71 -55.65 0.00
B120CXM__breadcrumb___5: 52.43 -59.19 0.00
A60W210LXM__breadcrumb___0: 87.56 -8.27 5.00
A60W210LXM__breadcrumb___1: 116.08 0.60 5.00
```

**Comment by caguero** [comment]
> > Do you have the poses of the breadcrumbs and robots by any chance?
> 
> The first simulation was `ver99octofqr1-81d7812a-7a4a-47cd-b7f8-b8f47c0a4370` with
> 
> ```
> A210LXM__breadcrumb___0: 29.54 0.57 2.90
> A210LXM__breadcrumb___1: 51.68 -19.57 5.00
> A210LXM__breadcrumb___2: 81.66 -18.82 5.00
> A210LXM__breadcrumb___3: 104.50 0.65 5.00
> A210LXM__breadcrumb___4: 134.22 0.62 0.67
> ```
> 
> The second simulation was `ver99octofqbread1-3e9df022-e221-45de-8a5f-a67f8821920f` with
> 
> ```
> B120CXM__breadcrumb___0: 29.54 -0.18 2.90
> B120CXM__breadcrumb___1: 48.18 -23.50 5.00
> B120CXM__breadcrumb___2: 47.93 -43.55 0.00
> B120CXM__breadcrumb___3: 47.81 -49.98 0.00
> B120CXM__breadcrumb___4: 48.71 -55.65 0.00
> B120CXM__breadcrumb___5: 52.43 -59.19 0.00
> A60W210LXM__breadcrumb___0: 87.56 -8.27 5.00
> A60W210LXM__breadcrumb___1: 116.08 0.60 5.00
> ```

This is great, could you also get the robot poses? This is critical to reproduce your experiment.

**Comment by caguero** [comment]
> > > Do you have the poses of the breadcrumbs and robots by any chance?
> > 
> > 
> > The first simulation was `ver99octofqr1-81d7812a-7a4a-47cd-b7f8-b8f47c0a4370` with
> > ```
> > A210LXM__breadcrumb___0: 29.54 0.57 2.90
> > A210LXM__breadcrumb___1: 51.68 -19.57 5.00
> > A210LXM__breadcrumb___2: 81.66 -18.82 5.00
> > A210LXM__breadcrumb___3: 104.50 0.65 5.00
> > A210LXM__breadcrumb___4: 134.22 0.62 0.67
> > ```
> > 
> > 
> > The second simulation was `ver99octofqbread1-3e9df022-e221-45de-8a5f-a67f8821920f` with
> > ```
> > B120CXM__breadcrumb___0: 29.54 -0.18 2.90
> > B120CXM__breadcrumb___1: 48.18 -23.50 5.00
> > B120CXM__breadcrumb___2: 47.93 -43.55 0.00
> > B120CXM__breadcrumb___3: 47.81 -49.98 0.00
> > B120CXM__breadcrumb___4: 48.71 -55.65 0.00
> > B120CXM__breadcrumb___5: 52.43 -59.19 0.00
> > A60W210LXM__breadcrumb___0: 87.56 -8.27 5.00
> > A60W210LXM__breadcrumb___1: 116.08 0.60 5.00
> > ```
> 
> This is great, could you also get the robot poses at the moment where the comms are strange? This is critical to reproduce your experiment.

**Comment by m3d** [comment]
Well the robots are moving and they are reporting its trajectory. And in the second case you can see that A robot waited 60s and then followed the **left** wall. The common part was placed by robot B, starting immediately and moving in the center (and later falling down from vertical shaft.
![bread-debug-Bview](https://user-images.githubusercontent.com/5664353/110624829-021a6700-819f-11eb-96a1-1c270cc97767.png)
And A robot is no longer visible from the Teambase.

**Comment by m3d** [comment]
p.s. maybe just checking - the breadcrumbs can be shared via all robots, right?

**Comment by caguero** [comment]
When a robot drops a breadcrumb, it affects all robots and breadcrumbs in its vicinity. Is that what you mean?

**Comment by m3d** [comment]
OK, I just wanted to check that really two breadcrumbs dropped by orange robot B should help also the blue robot A ... OK, then I still think that there is some bug and that blue robot radio range ends like if there are no breadcrumbs from orange robot.

**Comment by caguero** [comment]
@m3d , I've been studying your logs and the results that you're observing are mainly due to two reasons:

1. The breadcrumb location. To give you an example, see the picture below. I highlighted with yellow circles, two breadcrumbs that can't communicate each other because they're not close enough. It's right at the threshold (represented with a red line) but far enough to create a disconnected segment.

![team_log_highlights](https://user-images.githubusercontent.com/1440739/113929888-dc5d9d80-97f0-11eb-9300-d2ef75e02300.png)

2. The fact that tunnel tiles are small compared with other types (e.g. urban). Every time that we traverse a tile, we introduce a visibility cost. Therefore, when the tiles are small, the density of breadcrumbs need to be higher.

I just created pull request #884 to mitigate (2). The coverage should be better in tunnels with the same amount of breadcrumbs, making it closer to the other segment types.

My resolution is that this was expected behavior but we just created a patch to make the behavior more consistent in tunnels.
